File: typescript/tests/search.test.ts
import { Search } from '../src/resources/search';

describe('SERP search contract', () => {
  it('serializes defaults and new filters', async () => {
    const request = jest.fn().mockResolvedValue({ organic_results: [] });
    const search = new Search({ request } as any);

    await search.google({
      query: 'SDK examples',
      country: 'US',
      language: 'en',
      range: 'qdr:w',
      imageSize: '10mp',
    });

    expect(request).toHaveBeenCalledWith('POST', '/serp/google', {
      json: {
        query: 'SDK examples',
        type: 'search',
        page: 1,
        number: 10,
        country: 'US',
        language: 'en',
        range: 'qdr:w',
        image_size: '10mp',
      },
    });
  });

  it.each([
    [{ query: '   ' }, 'query'],
    [{ query: 'test', page: 0 }, 'page'],
    [{ query: 'test', number: 101 }, 'number'],
    [{ query: 'test', country: '' }, 'country'],
    [{ query: 'test', language: 'x'.repeat(33) }, 'language'],
    [{ query: 'test', range: 'invalid' } as any, 'range'],
    [{ query: 'test', imageSize: 'huge' } as any, 'imageSize'],
  ])('rejects invalid options', async (options, message) => {
    const request = jest.fn();
    const search = new Search({ request } as any);

    await expect(search.google(options)).rejects.toThrow(message);
    expect(request).not.toHaveBeenCalled();
  });
});

File: go/search_test.go
package acedatacloud

import "testing"

func TestSearchRequestSerializesDefaultsAndFilters(t *testing.T) {
	body, err := (SearchRequest{
		Query:     "SDK examples",
		Country:   "US",
		Language:  "en",
		Range:     SearchRangeQDRWeek,
		ImageSize: SearchImageSize10MP,
	}).toBody()
	if err != nil {
		t.Fatal(err)
	}

	if body["page"] != 1 || body["number"] != 10 || body["type"] != "search" {
		t.Fatalf("SERP defaults drifted: %#v", body)
	}
	if body["range"] != "qdr:w" || body["image_size"] != "10mp" {
		t.Fatalf("SERP filters drifted: %#v", body)
	}
}

func TestSearchRequestRejectsInvalidValues(t *testing.T) {
	tests := []SearchRequest{
		{Query: "   "},
		{Query: "test", Page: -1},
		{Query: "test", Number: 101},
		{Query: "test", Language: "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"},
		{Query: "test", Range: SearchRange("invalid")},
		{Query: "test", ImageSize: SearchImageSize("huge")},
	}

	for _, req := range tests {
		if _, err := req.toBody(); err == nil {
			t.Fatalf("expected validation error for %#v", req)
		}
	}
}
